core/src/test/scala/unit/kafka/controller/ControllerFailoverTest.scala

```diff
@@ -54,119 +53,40 @@ class ControllerFailoverTest extends KafkaServerTestHarness with Logging {
-      Thread.sleep(100)
-    }
-    // Replace channel manager with our mock manager
-    controller.kafkaController.controllerContext.controllerChannelManager.shutdown()
-    val channelManager = new MockChannelManager(controller.kafkaController.controllerContext, 
-                                                  controller.kafkaController.config, metrics)
-    channelManager.startup()
-    controller.kafkaController.controllerContext.controllerChannelManager = channelManager
-    channelManager.shrinkBlockingQueue(0)
-    channelManager.stopSendThread(0)
-    // Spawn a new thread to block on the outgoing channel
-    // queue
-    val thread = new Thread(new Runnable {
-      def run() {
-        try {
-          controller.kafkaController.sendUpdateMetadataRequest(Seq(0), Set(topicPartition))
-          log.info("Queue state %d %d".format(channelManager.queueCapacity(0), channelManager.queueSize(0)))
-          controller.kafkaController.sendUpdateMetadataRequest(Seq(0), Set(topicPartition))
-          log.info("Queue state %d %d".format(channelManager.queueCapacity(0), channelManager.queueSize(0)))
-        } catch {
-          case _: Exception => log.info("Thread interrupted")
-        }
-      }
+    TestUtils.waitUntilTrue(() =>
+      initialController.partitionStateMachine.partitionsInState(OnlinePartition).contains(topicPartition),
+      s"Partition $topicPartition did not transition to online state")
+
+    // Wait until we have verified that we have resigned
+    val latch = new CountDownLatch(1)
+    val illegalStateEvent = ControllerTestUtils.createMockControllerEvent(ControllerState.BrokerChange, { () =>
+      initialController.handleIllegalState(new IllegalStateException("Thrown for test purposes"))
+      latch.await()
     })
-    thread.setName("mythread")
-    thread.start()
-    while (thread.getState() != Thread.State.WAITING) {
-      Thread.sleep(100)
-    }
-    // Assume that the thread is WAITING because it is
-    // blocked on the queue, so interrupt and move forward
-    thread.interrupt()
-    thread.join()
-    channelManager.resumeSendThread(0)
-    // Wait and find current controller
-    var found = false
-    var counter = 0
-    while (!found && counter < 10) {
-      for (server <- this.servers) {
-        val previousEpoch = epochMap get server.config.brokerId match {
-          case Some(epoch) =>
-            epoch
-          case None =>
-            val msg = String.format("Missing element in epoch map %s", epochMap.mkString(", "))
-            throw new IllegalStateException(msg)
-        }
+    initialController.eventManager.put(illegalStateEvent)
+    // Check that we have shutdown the scheduler (via onControllerResigned)
+    TestUtils.waitUntilTrue(() => !initialController.kafkaScheduler.isStarted, "Scheduler was not shutdown")
+    assertEquals(None, zkUtils.readDataMaybeNull(ZkUtils.ControllerPath))
+    assertEquals(-1, initialController.getControllerID())
```

Comment:
Both of these should be `waitUntilTrue` as they happen after `onControllerResigned`